Lib/ntpath.py

```diff
@@ -554,52 +554,43 @@ def normpath(path):
         return prefix + sep.join(comps)
 
 
+def _abspath_fallback(path):
+    """Return the absolute version of a path as a fallback function in case
+    `nt._getfullpathname` is not available or raises OSError. See bpo-31047 for
+    more.
+
+    """
+
+    path = os.fspath(path)
+    if not isabs(path):
+        if isinstance(path, bytes):
+            sep = b'/'
+            cwd = os.getcwdb()
+        else:
+            sep = '/'
+            cwd = os.getcwd()
+        drive, root, path = splitroot(path)
+        if drive and drive != splitroot(cwd)[0]:
+            cwd = join(drive, sep)
+        path = join(cwd, root + path)
```

Comment:
I mean, per definition. Is there something unexpected here?

```python
>>> from ntpath import abspath
>>> abspath('C:/foo')
'C:\\foo'
>>> abspath('C:foo')
'C:\\foo'
>>> abspath('/foo')
'\\foo' # Yes, no drive...
>>> abspath('foo')
'\\Users\\wannes\\Documents\\GitHub\\cpython\\speedup-posixpath.abspath\\foo'
```

Reply:
Returning a relative path such as "\\foo" is nonsense because `isabs(abspath(p))` will be false. However, using `ntpath.abspath()` on POSIX has always been implemented nonsensically like this. This PR fixes the drive-relative case, but not the cases of simple relative and root relative paths.

Reply:
@barneygale, so for drive-relative paths, do you think the fallback code on POSIX should be reverted to using a naive join with the current working directory? That would preserve the existing bug.

The proper fallback code on Windows -- for when the path is too long or contains null characters -- would actually need to try to get the full path on the drive via `_path_abspath()`, instead of just assuming that it's the root directory.

Reply:
Reverted, then the fallback code on Windows can be more accurate.